In [ ]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from ast import literal_eval
from sklearn.feature_extraction.text import  CountVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.corpus import wordnet
from surprise import Reader, Dataset, SVD
from surprise.model_selection import train_test_split
from surprise import accuracy

import warnings; warnings.simplefilter('ignore')

Simple Recommender¶
The Simple Recommender offers generalized recommnendations to every user based on movie popularity and (sometimes) genre. 

In [2]:
md = pd. read_csv('/kaggle/input/the-movies-dataset/movies_metadata.csv')
md.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [3]:
md['genres'] = md['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

In [4]:
md

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                             genres                              homepage  \
0       [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   
1      [Adventure, Fantasy, Family]                                   NaN   
2                 [Romance, Comedy]                                   NaN   
3          [Comedy, Drama, Romance]                                   NaN   
4                          [Comedy]                                   NaN   
...                             ...                                   ...   
45461               [Drama, Family]  http://www.imdb.com/title/tt6209470/   
45462                       [Drama]                                   NaN   
45463     [Action, Drama, Thriller]                                   NaN   
45464                            []                                   NaN   
45465                            []                                   NaN   

           id    imdb_id original_language               original_title  \
0         862  tt0114709                en                    Toy Story   
1        8844  tt0113497                en                      Jumanji   
2       15602  tt0113228                en             Grumpier Old Men   
3       31357  tt0114885                en            Waiting to Exhale   
4       11862  tt0113041                en  Father of the Bride Part II   
...       ...        ...               ...                          ...   
45461  439050  tt6209470                fa                      رگ خواب   
45462  111109  tt2028550                tl          Siglo ng Pagluluwal   
45463   67758  tt0303758                en                     Betrayal   
45464  227506  tt0008536                en          Satana likuyushchiy   
45465  461257  tt6980792                en                     Queerama   

                                                overview  ... release_date  \
0      Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1      When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2      A family wedding reignites the ancient feud be...  ...   1995-12-22   
3      Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4      Just when George Banks has recovered from his ...  ...   1995-02-10   
...                                                  ...  ...          ...   
45461        Rising and falling between a man and woman.  ...          NaN   
45462  An artist struggles to finish his work while a...  ...   2011-11-17   
45463  When one of her hits goes wrong, a professiona...  ...   2003-08-01   
45464  In a small town live two brothers, one a minis...  ...   1917-10-21   
45465  50 years after decriminalisation of homosexual...  ...   2017-06-09   

           revenue runtime                                   spoken_languages  \
0      373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1      262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2              0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}

I use the TMDB Ratings to come up with our Top Movies Chart. I will use IMDB's weighted rating formula to construct my chart. Mathematically, it is represented as follows:

Weighted Rating (WR) = (v ÷ (v+m)) × R + (m ÷ (v+m)) × C 
 
where,

* v is the number of votes for the movie,
* m is the minimum votes required to be listed in the chart,
* R is the average rating of the movie
* C is the mean vote across the whole report

In [5]:
vote_counts = md[md['vote_count'].notnull()]['vote_count'].astype('int')
vote_averages = md[md['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_averages.mean()
C

5.244896612406511

In [6]:
m = vote_counts.quantile(0.95)
m

434.0

In [7]:
md['year'] = pd.to_datetime(md['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

In [8]:
qualified = md[(md['vote_count'] >= m) & (md['vote_count'].notnull()) & (md['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'genres']]
qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')
qualified.shape

(2274, 6)

To be eligible for inclusion in the chart, a movie must have received at least 434 votes on TMDB. Additionally, the average rating for a movie on TMDB is 5.244 out of 10. Based on these criteria, 2274 movies qualify to be featured on the chart

In [9]:
def weighted_rating(x):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [10]:
qualified['wr'] = qualified.apply(weighted_rating, axis=1)
qualified = qualified.sort_values('wr', ascending=False).head(250)


In [11]:
qualified.head(15)


title  year  vote_count  \
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
22879                                       Interstellar  2014       11187   
2843                                          Fight Club  1999        9678   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   
292                                         Pulp Fiction  1994        8670   
314                             The Shawshank Redemption  1994        8358   
7000       The Lord of the Rings: The Return of the King  2003        8226   
351                                         Forrest Gump  1994        8147   
5814               The Lord of the Rings: The Two Towers  2002        7641   
256                                            Star Wars  1977        6778   
1225                                  Back to the Future  1985        6239   
834                                        The Godfather  1972        6024   
1154                             The Empire Strikes Back  1980        5998   
46                                                 Se7en  1995        5915   

       vote_average  popularity  \
15480             8   29.108149   
12481             8  123.167259   
22879             8   32.213481   
2843              8   63.869599   
4863              8   32.070725   
292               8  140.950236   
314               8   51.645403   
7000              8   29.324358   
351               8   48.307194   
5814              8   29.423537   
256               8   42.149697   
1225              8   25.778509   
834               8   41.109264   
1154              8   19.470959   
46                8    18.45743   

                                                  genres        wr  
15480  [Action, Thriller, Science Fiction, Mystery, A...  7.917588  
12481                   [Drama, Action, Crime, Thriller]  7.905871  
22879                [Adventure, Drama, Science Fiction]  7.897107  
2843                                             [Drama]  7.881753  
4863                        [Adventure, Fantasy, Action]  7.871787  
292                                    [Thriller, Crime]  7.868660  
314                                       [Drama, Crime]  7.864000  
7000                        [Adventure, Fantasy, Action]  7.861927  
351                             [Comedy, Drama, Romance]  7.860656  
5814                        [Adventure, Fantasy, Action]  7.851924  
256                 [Adventure, Action, Science Fiction]  7.834205  
1225        [Adventure, Comedy, Science Fiction, Family]  7.820813  
834                                       [Drama, Crime]  7.814847  
1154                [Adventure, Action, Science Fiction]  7.814099  
46                            [Crime, Mystery, Thriller]  7.811669

In [12]:
s = md.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'genre'
gen_md = md.drop('genres', axis=1).join(s)

In [13]:
gen_md

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
1      False                                                NaN  65000000   
...      ...                                                ...       ...   
45463  False                                                NaN         0   
45463  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                                   homepage      id    imdb_id  \
0      http://toystory.disney.com/toy-story     862  tt0114709   
0      http://toystory.disney.com/toy-story     862  tt0114709   
0      http://toystory.disney.com/toy-story     862  tt0114709   
1                                       NaN    8844  tt0113497   
1                                       NaN    8844  tt0113497   
...                                     ...     ...        ...   
45463                                   NaN   67758  tt0303758   
45463                                   NaN   67758  tt0303758   
45463                                   NaN   67758  tt0303758   
45464                                   NaN  227506  tt0008536   
45465                                   NaN  461257  tt6980792   

      original_language       original_title  \
0                    en            Toy Story   
0                    en            Toy Story   
0                    en            Toy Story   
1                    en              Jumanji   
1                    en              Jumanji   
...                 ...                  ...   
45463                en             Betrayal   
45463                en             Betrayal   
45463                en             Betrayal   
45464                en  Satana likuyushchiy   
45465                en             Queerama   

                                                overview popularity  ...  \
0      Led by Woody, Andy's toys live happily in his ...  21.946943  ...   
0      Led by Woody, Andy's toys live happily in his ...  21.946943  ...   
0      Led by Woody, Andy's toys live happily in his ...  21.946943  ...   
1      When siblings Judy and Peter discover an encha...  17.015539  ...   
1      When siblings Judy and Peter discover an encha...  17.015539  ...   
...                                                  ...        ...  ...   
45463  When one of her hits goes wrong, a professiona...   0.903007  ...   
45463  When one of her hits goes wrong, a professiona...   0.903007  ...   
45463  When one of her hits goes wrong, a professiona...   0.903007  ...   
45464  In a small town live two brothers, one a minis...   0.003503  ...   
45465  50 years after decriminalisation of homosexual...   0.163015  ...   

      runtime                                   spoken_languages    status  \
0        81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
0        81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
0        81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1       104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
1       104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
...       ...                                                ...       ...   
45463    90.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
45463    90.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
45463    90.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
45464    87.0      

In [14]:
def build_chart(genre, percentile=0.85):
    df = gen_md[gen_md['genre'] == genre]
    vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(percentile)
    
    qualified = df[(df['vote_count'] >= m) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity']]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    
    qualified['wr'] = qualified.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C), axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(250)
    
    return qualified

In [15]:
build_chart('Romance').head(15)


title  year  vote_count  vote_average popularity  \
10309  Dilwale Dulhania Le Jayenge  1995         661             9  34.457024   
351                   Forrest Gump  1994        8147             8  48.307194   
876                        Vertigo  1958        1162             8   18.20822   
40251                   Your Name.  2016        1030             8  34.461252   
883               Some Like It Hot  1959         835             8  11.845107   
1132               Cinema Paradiso  1988         834             8  14.177005   
19901                     Paperman  2012         734             8   7.198633   
37863                  Sing Street  2016         669             8  10.672862   
882                  The Apartment  1960         498             8  11.994281   
38718               The Handmaiden  2016         453             8  16.727405   
3189                   City Lights  1931         444             8  10.891524   
24886             The Way He Looks  2014         262             8   5.711274   
45437               In a Heartbeat  2017         146             8   20.82178   
1639                       Titanic  1997        7770             7   26.88907   
19731      Silver Linings Playbook  2012        4840             7  14.488111   

             wr  
10309  8.565285  
351    7.971357  
876    7.811667  
40251  7.789489  
883    7.745154  
1132   7.744878  
19901  7.713951  
37863  7.689483  
882    7.599317  
38718  7.566166  
3189   7.558867  
24886  7.331363  
45437  7.003959  
1639   6.981546  
19731  6.970581

In [16]:
build_chart('Action').head(15)


title  year  vote_count  \
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   
7000       The Lord of the Rings: The Return of the King  2003        8226   
5814               The Lord of the Rings: The Two Towers  2002        7641   
256                                            Star Wars  1977        6778   
1154                             The Empire Strikes Back  1980        5998   
4135                                            Scarface  1983        3017   
9430                                              Oldboy  2003        2000   
1910                                       Seven Samurai  1954         892   
43190                                   Band of Brothers  2001         725   
1215                                                   M  1931         465   
14551                                             Avatar  2009       12114   
17818                                       The Avengers  2012       12000   
26564                                           Deadpool  2016       11444   

       vote_average  popularity        wr  
15480             8   29.108149  7.955099  
12481             8  123.167259  7.948610  
4863              8   32.070725  7.929579  
7000              8   29.324358  7.924031  
5814              8   29.423537  7.918382  
256               8   42.149697  7.908327  
1154              8   19.470959  7.896841  
4135              8   11.299673  7.802046  
9430              8   10.616859  7.711649  
1910              8    15.01777  7.426145  
43190             8    7.903731  7.325485  
1215              8   12.752421  7.072073  
14551             7  185.070892  6.966363  
17818             7   89.887648  6.966049  
26564             7  187.860492  6.964431

 Simple recommendation system will gives the same recommendation to everyone, regardless of the user's personal taste. If a person who loves romantic movies (and hates action) were to look at our Top 15 Chart, she/he wouldn't probably like most of the movies. If s/he were to go one step further and look at our charts by genre, s/he wouldn't still be getting the best recommendations.

# Content Based Recommender

It creates a user profile based on the features of the items the user has interacted with.
It recommends new items that share similar features with the ones the user has previously liked.

In [17]:
links = pd.read_csv('/kaggle/input/the-movies-dataset/links_small.csv')
links = links[links['tmdbId'].notnull()]['tmdbId'].astype('int')

In [18]:
def convert_int(x):
    try:
        return int(x)
    except:
        return np.nan

In [19]:
md['id'] = md['id'].apply(convert_int)


In [20]:
md[md['id'].isnull()]




adult  \
19730                                 - Written by Ørnås   
29503   Rune Balot goes to a casino connected to the ...   
35587   Avalanche Sharks tells the story of a bikini ...   

      belongs_to_collection                            budget  \
19730              0.065736  /ff9qCepilowshEtG2GYWwzt2bs4.jpg   
29503              1.931659  /zV8bHuSL6WXoD6FWogP9j4x80bL.jpg   
35587              2.185485  /zaSf5OG7V8X8gqFvly88zDdRm46.jpg   

                                                  genres  \
19730  [Carousel Productions, Vision View Entertainme...   
29503  [Aniplex, GoHands, BROSTA TV, Mardock Scramble...   
35587  [Odyssey Media, Pulser Productions, Rogue Stat...   

                                                homepage  id imdb_id  \
19730  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso... NaN       0   
29503  [{'iso_3166_1': 'US', 'name': 'United States o... NaN       0   
35587           [{'iso_3166_1': 'CA', 'name': 'Canada'}] NaN       0   

      original_language                            original_title  overview  \
19730             104.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   
29503              68.0      [{'iso_639_1': 'ja', 'name': '日本語'}]  Released   
35587              82.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   

       ... revenue runtime spoken_languages status tagline  title  video  \
19730  ...     NaN     NaN              NaN    NaN     NaN    NaN    NaN   
29503  ...     NaN     NaN              NaN    NaN     NaN    NaN    NaN   
35587  ...     NaN     NaN              NaN    NaN     NaN    NaN    NaN   

      vote_average vote_count year  
19730          NaN        NaN  NaT  
29503          NaN        NaN  NaT  
35587          NaN        NaN  NaT  

[3 rows x 25 columns]

In [21]:
md = md.drop([19730, 29503, 35587])
md['id'] = md['id'].astype('int')

In [22]:
smd = md[md['id'].isin(links)]
smd.shape

(9099, 25)

In [23]:
smd

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
40224  False                                                NaN  15000000   
40503  False                                                NaN         0   
44821  False  {'id': 34055, 'name': 'Pokémon Collection', 'p...  16000000   
44826  False  {'id': 34055, 'name': 'Pokémon Collection', 'p...         0   
45265  False                                                NaN         0   

                                                  genres  \
0                            [Animation, Comedy, Family]   
1                           [Adventure, Fantasy, Family]   
2                                      [Romance, Comedy]   
3                               [Comedy, Drama, Romance]   
4                                               [Comedy]   
...                                                  ...   
40224  [Action, Adventure, Drama, Horror, Science Fic...   
40503                               [Documentary, Music]   
44821    [Adventure, Fantasy, Animation, Action, Family]   
44826  [Adventure, Fantasy, Animation, Science Fictio...   
45265                                    [Comedy, Drama]   

                                                homepage      id    imdb_id  \
0                   http://toystory.disney.com/toy-story     862  tt0114709   
1                                                    NaN    8844  tt0113497   
2                                                    NaN   15602  tt0113228   
3                                                    NaN   31357  tt0114885   
4                                                    NaN   11862  tt0113041   
...                                                  ...     ...        ...   
40224                                                NaN  315011  tt4262980   
40503              http://www.thebeatlesliveproject.com/  391698  tt2531318   
44821                  http://movies.warnerbros.com/pk3/   10991  tt0235679   
44826  http://www.pokemon.com/us/movies/movie-pokemon...   12600  tt0287635   
45265                                                NaN  265189  tt2121382   

      original_language                                     original_title  \
0                    en                                          Toy Story   
1                    en                                            Jumanji   
2                    en                                   Grumpier Old Men   
3                    en                                  Waiting to Exhale   
4                    en                        Father of the Bride Part II   
...                 ...                                                ...   
40224                ja                                             シン・ゴジラ   
40503                en  The Beatles: Eight Days a Week - The Touring Y...   
44821                ja                               Pokémon 3: The Movie   
44826                ja                     劇場版ポケットモンスター セレビィ 時を越えた遭遇（であい）   
45265                sv                                             Turist   

                                                overview  ...      revenue  \
0      Led by Woody, Andy's toys live happily in his ...  ...  373554033.0   
1      When siblings Judy and Peter discover an encha...  ...  262797249.0   
2      A family wedding reignites the ancient feud be...  ...          0.0   
3      Cheated on, mistreated and stepped on, the wom...  ...   81452156.0   
4      Just when George Banks has recovered from his ...  ...   76578911.0   
...    

In [24]:
smd['tagline'] = smd['tagline'].fillna('')
smd['description'] = smd['overview'] + smd['tagline']
smd['description'] = smd['description'].fillna('')

In [25]:
credits = pd.read_csv('/kaggle/input/the-movies-dataset/credits.csv')
keywords = pd.read_csv('/kaggle/input/the-movies-dataset/keywords.csv')

In [26]:
keywords['id'] = keywords['id'].astype('int')
credits['id'] = credits['id'].astype('int')
md['id'] = md['id'].astype('int')

In [27]:
md.shape


(45463, 25)

In [28]:
md = md.merge(credits, on='id')


In [29]:
md = md.merge(keywords, on='id')


In [30]:
md

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
46623  False                                                NaN         0   
46624  False                                                NaN         0   
46625  False                                                NaN         0   
46626  False                                                NaN         0   
46627  False                                                NaN         0   

                             genres                              homepage  \
0       [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   
1      [Adventure, Fantasy, Family]                                   NaN   
2                 [Romance, Comedy]                                   NaN   
3          [Comedy, Drama, Romance]                                   NaN   
4                          [Comedy]                                   NaN   
...                             ...                                   ...   
46623               [Drama, Family]  http://www.imdb.com/title/tt6209470/   
46624                       [Drama]                                   NaN   
46625     [Action, Drama, Thriller]                                   NaN   
46626                            []                                   NaN   
46627                            []                                   NaN   

           id    imdb_id original_language               original_title  \
0         862  tt0114709                en                    Toy Story   
1        8844  tt0113497                en                      Jumanji   
2       15602  tt0113228                en             Grumpier Old Men   
3       31357  tt0114885                en            Waiting to Exhale   
4       11862  tt0113041                en  Father of the Bride Part II   
...       ...        ...               ...                          ...   
46623  439050  tt6209470                fa                      رگ خواب   
46624  111109  tt2028550                tl          Siglo ng Pagluluwal   
46625   67758  tt0303758                en                     Betrayal   
46626  227506  tt0008536                en          Satana likuyushchiy   
46627  461257  tt6980792                en                     Queerama   

                                                overview  ...    status  \
0      Led by Woody, Andy's toys live happily in his ...  ...  Released   
1      When siblings Judy and Peter discover an encha...  ...  Released   
2      A family wedding reignites the ancient feud be...  ...  Released   
3      Cheated on, mistreated and stepped on, the wom...  ...  Released   
4      Just when George Banks has recovered from his ...  ...  Released   
...                                                  ...  ...       ...   
46623        Rising and falling between a man and woman.  ...  Released   
46624  An artist struggles to finish his work while a...  ...  Released   
46625  When one of her hits goes wrong, a professiona...  ...  Released   
46626  In a small town live two brothers, one a minis...  ...  Released   
46627  50 years after decriminalisation of homosexual...  ...  Released   

                                                 tagline  \
0                                                    NaN   
1              Roll the dice and unleash the excitement!   
2      Still Yelling. Still Fighting. Still Ready for...   
3      Friends are the people who let you be yourself...   
4      Just When His World Is Back To Normal... He's ..

In [31]:
smd = md[md['id'].isin(links)]
smd.shape


(9219, 28)

In [32]:
smd.columns

Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count', 'year', 'cast', 'crew', 'keywords'],
      dtype='object')

In [33]:
smd['cast'][0]

"[{'cast_id': 14, 'character': 'Woody (voice)', 'credit_id': '52fe4284c3a36847f8024f95', 'gender': 2, 'id': 31, 'name': 'Tom Hanks', 'order': 0, 'profile_path': '/pQFoyx7rp09CJTAb932F2g8Nlho.jpg'}, {'cast_id': 15, 'character': 'Buzz Lightyear (voice)', 'credit_id': '52fe4284c3a36847f8024f99', 'gender': 2, 'id': 12898, 'name': 'Tim Allen', 'order': 1, 'profile_path': '/uX2xVf6pMmPepxnvFWyBtjexzgY.jpg'}, {'cast_id': 16, 'character': 'Mr. Potato Head (voice)', 'credit_id': '52fe4284c3a36847f8024f9d', 'gender': 2, 'id': 7167, 'name': 'Don Rickles', 'order': 2, 'profile_path': '/h5BcaDMPRVLHLDzbQavec4xfSdt.jpg'}, {'cast_id': 17, 'character': 'Slinky Dog (voice)', 'credit_id': '52fe4284c3a36847f8024fa1', 'gender': 2, 'id': 12899, 'name': 'Jim Varney', 'order': 3, 'profile_path': '/eIo2jVVXYgjDtaHoF19Ll9vtW7h.jpg'}, {'cast_id': 18, 'character': 'Rex (voice)', 'credit_id': '52fe4284c3a36847f8024fa5', 'gender': 2, 'id': 12900, 'name': 'Wallace Shawn', 'order': 4, 'profile_path': '/oGE6JqPP2xH4t

In [34]:
smd['cast'] = smd['cast'].apply(literal_eval)
smd['crew'] = smd['crew'].apply(literal_eval)
smd['keywords'] = smd['keywords'].apply(literal_eval)
smd['cast_size'] = smd['cast'].apply(lambda x: len(x))
smd['crew_size'] = smd['crew'].apply(lambda x: len(x))

In [35]:
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

In [36]:
smd['director'] = smd['crew'].apply(get_director)


In [37]:
smd['cast'] = smd['cast'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])


In [38]:
smd['cast'] = smd['cast'].apply(lambda x: x[:4] if len(x) >=4 else x)

In [39]:
smd['cast']

0          [Tom Hanks, Tim Allen, Don Rickles, Jim Varney]
1        [Robin Williams, Jonathan Hyde, Kirsten Dunst,...
2        [Walter Matthau, Jack Lemmon, Ann-Margret, Sop...
3        [Whitney Houston, Angela Bassett, Loretta Devi...
4        [Steve Martin, Diane Keaton, Martin Short, Kim...
                               ...                        
45985    [Veronica Taylor, Rachael Lillis, Maddie Blaus...
46424    [Lisa Loven Kongsli, Johannes Bah Kuhnke, Clar...
46425    [Lisa Loven Kongsli, Johannes Bah Kuhnke, Clar...
46426    [Lisa Loven Kongsli, Johannes Bah Kuhnke, Clar...
46427    [Lisa Loven Kongsli, Johannes Bah Kuhnke, Clar...
Name: cast, Length: 9219, dtype: object

In [40]:
smd['keywords'] = smd['keywords'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

In [41]:
smd['cast'] = smd['cast'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])


In [42]:
smd['director'] = smd['director'].astype('str').apply(lambda x: str.lower(x.replace(" ", "")))
smd['director'] = smd['director'].apply(lambda x: [x,x, x])

In [43]:
s = smd.apply(lambda x: pd.Series(x['keywords']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'keyword'

In [44]:
s = s.value_counts()
s[:5]

keyword
independent film        610
woman director          550
murder                  399
duringcreditsstinger    327
based on novel          318
Name: count, dtype: int64

In [45]:
s = s[s > 1]

In [46]:
stemmer = SnowballStemmer('english')
stemmer.stem('dogs')

'dog'

In [47]:
def filter_keywords(x):
    words = []
    for i in x:
        if i in s:
            words.append(i)
    return words

In [48]:
md['keywords'] = smd['keywords'].apply(filter_keywords)
smd['keywords'] = smd['keywords'].apply(lambda x: [stemmer.stem(i) for i in x])
smd['keywords'] = smd['keywords'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [49]:
smd.columns

Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count', 'year', 'cast', 'crew', 'keywords',
       'cast_size', 'crew_size', 'director'],
      dtype='object')

In [50]:
smd['soup'] = smd['keywords'] + smd['cast'] + smd['director'] + smd['genres']
smd['soup'] = smd['soup'].apply(lambda x: ' '.join(x))

In [51]:
smd['soup'][0]

'jealousi toy boy friendship friend rivalri boynextdoor newtoy toycomestolif tomhanks timallen donrickles jimvarney johnlasseter johnlasseter johnlasseter Animation Comedy Family'

In [52]:
count = CountVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
count_matrix = count.fit_transform(smd['soup'])

In [53]:
count_matrix

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 271141 stored elements and shape (9219, 132041)>

In [54]:
cosine_sim = cosine_similarity(count_matrix, count_matrix)

In [55]:
cosine_sim

array([[1.        , 0.02328101, 0.02519763, ..., 0.0204765 , 0.0204765 ,
        0.0204765 ],
       [0.02328101, 1.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.02519763, 0.        , 1.        , ..., 0.02321817, 0.02321817,
        0.02321817],
       ...,
       [0.0204765 , 0.        , 0.02321817, ..., 1.        , 1.        ,
        1.        ],
       [0.0204765 , 0.        , 0.02321817, ..., 1.        , 1.        ,
        1.        ],
       [0.0204765 , 0.        , 0.02321817, ..., 1.        , 1.        ,
        1.        ]])

In [56]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])

In [57]:
def get_recommendations(title, cosine_sim=cosine_sim):
    # Check if the title is in the DataFrame
    if title not in smd['title'].values:
        raise ValueError("The movie title is not in the DataFrame")

    # Find the index of the movie with the given title
    idx = smd.index[smd['title'] == title].tolist()[0]
    
    # Check if the index is valid
    if idx >= len(cosine_sim):
        raise IndexError("Index out of range in similarity matrix")
    
    # Get the pairwise similarity scores of all movies with that movie
    sim_scores = list(enumerate(cosine_sim[idx]))

    # Ensure sim_scores contains valid tuples (index, score)
    if not all(isinstance(x, tuple) and len(x) == 2 for x in sim_scores):
        raise ValueError("Similarity scores are not in the expected format")

    # Sort the movies based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the top 30 most similar movies (excluding itself)
    sim_scores = sim_scores[1:31]  # Adjust as needed
    movie_indices = [i[0] for i in sim_scores]

    movies = smd.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year']]
    vote_counts = movies[movies['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = movies[movies['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(0.60)
    qualified = movies[(movies['vote_count'] >= m) & (movies['vote_count'].notnull()) & (movies['vote_average'].notnull())]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    qualified['wr'] = qualified.apply(weighted_rating, axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(10)

    return qualified



In [58]:
print(get_recommendations('The Dark Knight'))

                      title  vote_count  vote_average  year        wr
7608              Inception       14075             8  2010  7.917588
8573           Interstellar       11187             8  2014  7.897107
6587           The Prestige        4510             8  2006  7.758148
3373                Memento        4168             8  2000  7.740175
7991  The Dark Knight Rises        9263             7  2012  6.921448
6186          Batman Begins        7511             7  2005  6.904127
1122         Batman Returns        1706             6  1992  5.846862
132          Batman Forever        1529             5  1995  5.054144
9151      London Has Fallen        1656             5  2016  5.050854
9150      London Has Fallen        1656             5  2016  5.050854


In [59]:
print(get_recommendations('Memento'))

                      title  vote_count  vote_average  year        wr
7608              Inception       14075             8  2010  7.917588
6945        The Dark Knight       12269             8  2008  7.905871
8573           Interstellar       11187             8  2014  7.897107
6587           The Prestige        4510             8  2006  7.758148
7991  The Dark Knight Rises        9263             7  2012  6.921448
6186          Batman Begins        7511             7  2005  6.904127
4153        Minority Report        2663             7  2002  6.754048
1011              Chinatown         939             7  1974  6.445219
2077              Following         363             7  1998  6.044272
4125               Insomnia        1181             6  2002  5.797081


In [60]:
print(get_recommendations('Star Wars'))

                                             title  vote_count  vote_average  \
958                        The Empire Strikes Back        5998             8   
8833                  Star Wars: The Force Awakens        7993             7   
8840                    Captain America: Civil War        7462             7   
971                             Return of the Jedi        4763             7   
6167  Star Wars: Episode III - Revenge of the Sith        4200             7   
7560                                    Iron Man 2        6969             6   
8541                                     Divergent        4784             6   
2112     Star Wars: Episode I - The Phantom Menace        4526             6   
3041                                         X-Men        4172             6   
4117  Star Wars: Episode II - Attack of the Clones        4074             6   

      year        wr  
958   1980  7.814099  
8833  2015  6.909610  
8840  2016  6.903532  
971   1983  6.853432  
6167

# Collabarative Recommendation

In [61]:
reader = Reader()
ratings = pd.read_csv('/kaggle/input/the-movies-dataset/ratings_small.csv')
ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [140]:
# Define the rating scale
reader = Reader(rating_scale=(0.5, 5))

# Load the data

data = Dataset.load_from_df(ratings[['userId', 'movieId', 'rating']], reader)

# Split the data into training and test sets
trainset, testset = train_test_split(data, test_size=0.2)

# Initialize the SVD model
svd = SVD()

# Train the model
svd.fit(trainset)

# Evaluate the model
predictions = svd.test(testset)
accuracy.rmse(predictions)
accuracy.mae(predictions)


RMSE: 0.8998
MAE:  0.6874


0.6874371231572369

In [141]:
ratings[ratings['userId'] == 1]

userId  movieId  rating   timestamp
0        1       31     2.5  1260759144
1        1     1029     3.0  1260759179
2        1     1061     3.0  1260759182
3        1     1129     2.0  1260759185
4        1     1172     4.0  1260759205
5        1     1263     2.0  1260759151
6        1     1287     2.0  1260759187
7        1     1293     2.0  1260759148
8        1     1339     3.5  1260759125
9        1     1343     2.0  1260759131
10       1     1371     2.5  1260759135
11       1     1405     1.0  1260759203
12       1     1953     4.0  1260759191
13       1     2105     4.0  1260759139
14       1     2150     3.0  1260759194
15       1     2193     2.0  1260759198
16       1     2294     2.0  1260759108
17       1     2455     2.5  1260759113
18       1     2968     1.0  1260759200
19       1     3671     3.0  1260759117

In [142]:
svd.predict(1, 2455)


Prediction(uid=1, iid=2455, r_ui=None, est=2.573871884099689, details={'was_impossible': False})

# Combination of content and collaborative recommendation. 

In [143]:
def convert_int(x):
    try:
        return int(x)
    except:
        return np.nan

In [144]:
id_map = pd.read_csv('/kaggle/input/the-movies-dataset/links_small.csv')[['movieId', 'tmdbId']]
id_map['tmdbId'] = id_map['tmdbId'].apply(convert_int)
id_map.columns = ['movieId', 'id']
id_map = id_map.merge(smd[['title', 'id']], on='id').set_index('title')

In [145]:
indices_map = id_map.set_index('id')


In [146]:
indices_map

movieId
id               
862.0           1
8844.0          2
15602.0         3
31357.0         4
11862.0         5
...           ...
159550.0   161944
392572.0   162542
402672.0   162672
315011.0   163056
391698.0   163949

[9219 rows x 1 columns]

In [154]:
def hybrid(userId, title):
    idx = indices[title]
    tmdbId = id_map.loc[title]['id']
    #print(idx)
    movie_id = id_map.loc[title]['movieId']
    
    sim_scores = list(enumerate(cosine_sim[int(idx)]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:26]
    movie_indices = [i[0] for i in sim_scores]
    
    movies = smd.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year', 'id']]
    movies['est'] = movies['id'].apply(lambda x: svd.predict(userId, indices_map.loc[x]['movieId']).est)
    movies = movies.sort_values('est', ascending=False)
    return movies.head(10)

In [155]:
hybrid(1, 'Avatar')



title  vote_count  vote_average  year      id  \
522       Terminator 2: Judgment Day      4274.0           7.7  1991     280   
999                   The Terminator      4208.0           7.4  1984     218   
2006                Fantastic Planet       140.0           7.6  1973   16306   
910                        The Abyss       822.0           7.1  1989    2756   
8622      X-Men: Days of Future Past      6155.0           7.5  2014  127585   
962                           Aliens      3282.0           7.7  1986     679   
8357         Star Trek Into Darkness      4479.0           7.4  2013   54138   
2826                        Predator      2129.0           7.3  1987     106   
4323  Piranha Part Two: The Spawning        41.0           3.9  1981   31646   
4001                 Hawk the Slayer        13.0           4.5  1980   25628   

           est  
522   3.384610  
999   3.112928  
2006  3.107932  
910   3.064520  
8622  2.984000  
962   2.973199  
8357  2.934055  
2826  2.921304  
4323  2.901441  
4001  2.856356

In [151]:
hybrid(500, 'Avatar')


title  vote_count  vote_average  year      id  \
522        Terminator 2: Judgment Day      4274.0           7.7  1991     280   
1368                          Titanic      7770.0           7.5  1997     597   
999                    The Terminator      4208.0           7.4  1984     218   
8357          Star Trek Into Darkness      4479.0           7.4  2013   54138   
4323   Piranha Part Two: The Spawning        41.0           3.9  1981   31646   
344                         True Lies      1138.0           6.8  1994   36955   
3052  Sinbad and the Eye of the Tiger        39.0           6.3  1977   11940   
8622       X-Men: Days of Future Past      6155.0           7.5  2014  127585   
2006                 Fantastic Planet       140.0           7.6  1973   16306   
910                         The Abyss       822.0           7.1  1989    2756   

           est  
522   3.615061  
1368  3.276856  
999   3.243019  
8357  3.176139  
4323  3.111979  
344   3.078452  
3052  3.029482  
8622  3.002236  
2006  2.996911  
910   2.967924